### PR TITLE
Episode 2 is updated with challenges... still working on that.

### PR DESCRIPTION
**Pull request by prithwish-ichec**
Please delete this line and the text below before submitting your contribution.

---

Thanks for contributing! If this contribution is for instructor training, please send an email to checkout@carpentries.org with a link to this contribution so we can record your progress. You’ve completed your contribution step for instructor checkout just by submitting this contribution.  

Please keep in mind that lesson maintainers are volunteers and it may be some time before they can respond to your contribution. Although not all contributions can be incorporated into the lesson materials, we appreciate your time and effort to improve the curriculum.  If you have any questions about the lesson maintenance process or would like to volunteer your time as a contribution reviewer, please contact The Carpentries Team at team@carpentries.org.  

---

### CODE REVIEW

File: _episodes/02-benchmark-and-scaling.md [hunk 1]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
 ## Case study: Benchmarking
 
-(FIXME)
+By now you already got some understanding about how benchmarking enables you to compare the performance of your computing system with some 'standard' systems, and thus it helps you to know whether a further fine tuning of your job is required or not. 
+As a first exercise, we will start with the simplest Lennard-Jones (LJ) system as provided in the 'bench' directory of the latest LAMMPS distribution. Following this, you will be given another example and you will follow the same set of exercises and compare your results with some 'known' benchmark.
+
+> ## Run a LAMMPS job in a HPC
+> Can you list bare minimum files that you need to schedule a LAMMPS job in a HPC?
+> > ## Solution
+> > For running a LAMMPS job, we must need an input file, empirical potential file (optional), and a data file (optional). We need these optional files only when the empirical potential model parameters and the molecular coordinates are not defined within the LAMMPS input file. In addition, for submitting a LAMMPS job to a HPC queue, we need a batch file. In general, each HPC uses a queueing system to manage all the jobs. Two such popular ones are PBS and Slurm.
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:38]: Maybe present answer in bullet point with a brief sentence afterwards instead of whole paragraph?

File: _config.yml [hunk 2]
```diff
@@ -60,6 +60,35 @@ lc_post_survey: "https://www.surveymonkey.com/r/lcpostworkshopsurvey?workshop_id
 instructor_pre_survey: "https://www.surveymonkey.com/r/instructor_training_pre_survey?workshop_id="
 instructor_post_survey: "https://www.surveymonkey.com/r/instructor_training_post_survey?workshop_id="
 
+# Workshop Specific Values
```
Comment (cilwerner) [inline _config.yml:64]: Work in progress, but {{ site.commands }} working effectively now.

File: _episodes/02-benchmark-and-scaling.md (hunk 1 above)
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:38]: Capital SLURM at sentence end.

File: _episodes/02-benchmark-and-scaling.md [hunk 3]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+```
+{: .Input}
+
+
+The timing information for this run with both 1 and 4 processors is also provided with LAMMPS distribution. So, to benchmark it would be wise to run the same job with same processor settings. Let us now create a batch file to submit this job. 
+
+
+> ## Batch file to submit a LAMMPS job
+>
+> Here is shown the header of a SLURM script to submit a LAMMPS job using 2 nodes and 48 processors. Can you modify the necessary fields in this script to submit the job using 4 processors?
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:54]: Another abstraction for SLURM would need to go in here depending on the system this is being run on, eg. {{ site.sched_name }} for SLURM or PBS? 

To be addressed also in config.yml

File: _episodes/02-benchmark-and-scaling.md [hunk 4]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+>
+> ~~~
+> #!/bin/bash -x
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:57]: `--nodes`, `--account` etc need abstractions

File: _episodes/02-benchmark-and-scaling.md [hunk 5]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+> > ## Solution
+> > ~~~
+> > {{ site.sched_comment }} --nodes=1
+> > {{ site.sched_comment }} --ntasks-per-node=4
+> > ~~~
+> > {: .input}
+> {: .solution}
+{: .challenge}
+
+Okay, the rest part of the batch file should now be used to set the LAMMPS environment variables and finally we need to invoke the LAMMPS executable for the run. In all modern HPCs, package specific environment variables are loaded or unloaded using modules. Module is a piece of code that is load or unloaded into Kernel as required thus eliminating the requirement of system reboot each time you need to extend the functionalities of the Kernel. Here we show an example code on how to use the 'module' command to load LAMMPS in Jureca.
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:80]: Instead of 'example code on how to use the 'module' command to load LAMMPS in Jureca.', this would likely need changing to an abstraction, as different presenters would be using different machines.

config.yml option of `{{ site.host_name }}`

File: _episodes/02-benchmark-and-scaling.md [hunk 6]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+
+~~~
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:82]: As above module load option needs abstraction.

File: _episodes/02-benchmark-and-scaling.md [hunk 7]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+module use /usr/local/software/jureca/OtherStages
+module load Stages/Devel-2019a
+module load intel-para/2019a
+module load LAMMPS/18Feb2020-cuda
+~~~
+{: .source}
+
+> ## Module
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:90]: Would this depend more on the machine? Too generic?

File: _episodes/02-benchmark-and-scaling.md [hunk 8]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+> In the above example, this is evident that ```LAMMPS/9Jan2020-cuda``` is the LAMMPS module that load the LAMMPS specific environment variables. Can you tell now why we loaded other modules like intel-para/2019a?
+>
+> > ## Solution
+> > LAMMPS was built using Intel compiler and could have dynamic linking set to some of the important libraries that are called at runtime.  
+> {: .solution}
 {: .challenge}
 
+Now we'll invoke LAMMPS executable to do the job. This can be done as following
+```
+srun lmp < in.lj
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:101]: Abstraction to be added.

File: _episodes/02-benchmark-and-scaling.md [hunk 9]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+> ## Module 
 >
+> In the above example, this is evident that ```LAMMPS/9Jan2020-cuda``` is the LAMMPS module that load the LAMMPS specific environment variables. Can you tell now why we loaded other modules like intel-para/2019a?
+>
+> > ## Solution
+> > LAMMPS was built using Intel compiler and could have dynamic linking set to some of the important libraries that are called at runtime.  
+> {: .solution}
 {: .challenge}
 
+Now we'll invoke LAMMPS executable to do the job. This can be done as following
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:99]: This may need rewording to fit with abstraction... or does it work as is?

File: _episodes/02-benchmark-and-scaling.md [hunk 10]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+```
+{: .input}
+
+In this case, ```lmp``` is the name of the LAMMPS executable. But, in your HPC it could named something else.
+
+> ## Slurm script: full view
+>
+> Just following the exercise above can you create a batch file to submit a LAMMPS job for the above input file (say, in.lj) to 1 core only. You will submit the job to a partition/queue named ```batch```. The job is expected to take not more than 5 minutes, and the ```batch``` partition allows you to submit jobs not crossing 72 hours time limit. The name of the LAMMPS executable is ```lmp```. 
+>
+> > ## Solution
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:117]: As above `--account` etc

File: _episodes/02-benchmark-and-scaling.md [hunk 11]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+```
+LAMMPS (18 Feb 2020)
+OMP_NUM_THREADS environment is not set. Defaulting to 1 thread. (src/comm.cpp:94)
+  using 1 OpenMP thread(s) per MPI task
+```
+Note that it tells you about the LAMMPS version, and ```OMP_NUM_THREADS``` which is one of the important environment variables required to set for obtainging a performance boost. This would be discussed later in this tutorial. But for now, we'll focus mainly on the timing information provided in this file. 
+
+When the run concludes, LAMMPS prints the final thermodynamic state and a total run time for the simulation. It also appends statistics about the CPU time and storage requirements for the simulation. An example set of statistics is shown here:
+
+```
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:152]: As snippet of code from `_includes` maybe?

File: _episodes/02-benchmark-and-scaling.md [hunk 12]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+  * The ```Loop time``` is the total wall-clock time for the simulation to run. (source: LAMMPS manual)
+Use the following to extract this from ```log.lammps```:
+```
+grep "Loop time" log.lammps| sed 's/|/ /' | awk '{print $4}'
+```
+```
+$ 1.76553
+```
+
+  * The Performance line is provided for convenience to help predict how long it will take to run a desired physical simulation.  (source: LAMMPS manual)
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:193]: add code marks to make `Performance` stand out

Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:193]: Reference LAMMPS manual with `[LAMMPS](link)`

File: _episodes/02-benchmark-and-scaling.md [hunk 13]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+Use the following command line to extract the value in units of ```tau/day``` :
+```
+grep "Performance" log.lammps| sed 's/|/ /' | awk '{print $2}'
+```
+```
+$ 24468.549
+```
+
+  * The CPU use line provides the CPU utilization per MPI task; it should be close to 100% times the number of OpenMP threads (or 1 of not using OpenMP). Lower numbers correspond to delays due to file I/O or insufficient thread utilization. (source: LAMMPS manual)
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:202]: Reference LAMMPS manual with `[LAMMPS](link)`

File: _episodes/02-benchmark-and-scaling.md [hunk 14]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+  * Comm = inter-processor communication of atoms and their properties
+  * Output = output of thermodynamic info and dump files
+  * Modify = fixes and computes invoked by fixes
+  * Other = all the remaining time
+
+This is very useful in the sense that it helps you to identify the performance bottlenecks! It will discussed later in more detail.
+
+> ## Now run a benchmark...
+>
+> Now submit a LAMMPS job for the above input file in a HPC that is available to you using both 1 and 4 processors. Extract the loop times for your runs and make a bar plot to see how the performance of your HPC for this particular job compares with LAMMPS standard benchmark and with the performance for another two HPCs, Kay and Jureca.
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:241]: Instead of calling them by name, should we just call them HPC1, HPC2?

File: _episodes/02-benchmark-and-scaling.md [hunk 15]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+```
+mpirun -np 1 lmp < in.lj
+```
+{: .input}
+
+In this case, ```lmp``` is the name of the LAMMPS executable. But, in your HPC it could named something else.
+
+> ## Slurm script: full view
+>
+> Just following the exercise above can you create a batch file to submit a LAMMPS job for the above input file (say, in.lj) to 1 core only. You will submit the job to a partition/queue named ```batch```. The job is expected to take not more than 5 minutes, and the ```batch``` partition allows you to submit jobs not crossing 72 hours time limit. The name of the LAMMPS executable is ```lmp```.
```
Comment (cilwerner) [inline _episodes/02-benchmark-and-scaling.md:115]: Add required abstractions for `sbatch` and `lmp`

File: _includes/snippets/ep02/runjob.sh [hunk 16]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash -x
```
Comment (cilwerner) [inline _includes/snippets/ep02/runjob.sh:1]: Is this file needed now? Mentioned twice in 02-benchmark-and-scaling. Or can `{{ site.stuff }}` be used in runjob.sh and be run as if it were PBS or SBATCH. Do the abstractions make that transition?

File: _episodes/02-benchmark-and-scaling.md [hunk 17]
```diff
@@ -19,27 +19,244 @@ keypoints:
 
 ## What is benchmarking?
 
-(FIXME)
+Benchmarking can be viewed like a sprint athelete. The athelete runs a predetermined distance on a particular surface, and a time is recorded. Based on different conditions, such as how dry or wet the surface is, or what the surface is made of, grass, sand, or track, the times of the sprinter to cover a distance (100m, 200m, 400m etc) will differ. If you had no idea where the sprinter was running, or what the conditions were like, if the sprinter sets a certain time you can cross-correlate it with a certain time associated with certain surfaces.
```
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:22]:
```
Now, when a sprinter runs a certain time in competition you can cross-correlate it with a times associated with those types of conditions and surfaces to judge how well they are performing.
```

File: _episodes/02-benchmark-and-scaling.md [hunk 18]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
 
 (FIXME)
 
 ## Case study: Benchmarking
 
-(FIXME)
+By now you already got some understanding about how benchmarking enables you to compare the performance of your computing system with some 'standard' systems, and thus it helps you to know whether a further fine tuning of your job is required or not. 
+As a first exercise, we will start with the simplest Lennard-Jones (LJ) system as provided in the 'bench' directory of the latest LAMMPS distribution. Following this, you will be given another example and you will follow the same set of exercises and compare your results with some 'known' benchmark.
+
+> ## Run a LAMMPS job in a HPC
```
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:35]: `What is needed to run a LAMMPS job on an HPC system?`

File: _episodes/02-benchmark-and-scaling.md (hunk 4 above)
Comment (ocaisa) [reply _episodes/02-benchmark-and-scaling.md:57]: So I wonder if we need to abstract every detail, it's going to be really hard to get this consistent across all possible schedulers.

We need to look at the overall requirement of each of our jobs, perhaps we can abstract things out in lumps (for example the bash shebang, the account, the partition, the time (?), the output/error files). We probably only want a couple of these big lumps to cover all cases. We should talk about this.

File: _episodes/02-benchmark-and-scaling.md (hunk 7 above)
Comment (ocaisa) [reply _episodes/02-benchmark-and-scaling.md:90]: I think I would probably use something generic in the template that gives you the environment to load LAMMPS. We shouldn't go into the description of what is inside that here, the instructor should do that themselves on the day for the machine that they are using

File: _episodes/02-benchmark-and-scaling.md [hunk 19]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+```
+{: .input}
+
+Or, if you are using ```mpirun``` instead of ```srun```, the submission line would be:
+```
+mpirun -np 1 lmp < in.lj
+```
+{: .input}
+
+In this case, ```lmp``` is the name of the LAMMPS executable. But, in your HPC it could named something else.
```
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:111]: Ah true, this would also need abstracting

File: _episodes/02-benchmark-and-scaling.md [hunk 20]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+
+Once you open the ```log.lammps``` file you will notice that this file contains most of the important information starting from the LAMMPS version, number of processors used for the runs, processor lay out, thermdynamic steps, and the timing information. The header of a ```log.lammps``` file would be somewhat similar to this:
+```
+LAMMPS (18 Feb 2020)
+OMP_NUM_THREADS environment is not set. Defaulting to 1 thread. (src/comm.cpp:94)
+  using 1 OpenMP thread(s) per MPI task
+```
+Note that it tells you about the LAMMPS version, and ```OMP_NUM_THREADS``` which is one of the important environment variables required to set for obtainging a performance boost. This would be discussed later in this tutorial. But for now, we'll focus mainly on the timing information provided in this file.
```
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:148]: `obtaining`

File: _episodes/02-benchmark-and-scaling.md (hunk 11 above)
Comment (ocaisa) [reply _episodes/02-benchmark-and-scaling.md:152]: Agree, this is better suited to a snippet so that it can be easily changed/updated.

File: _episodes/02-benchmark-and-scaling.md [hunk 21]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+Dangerous builds not checked
+Total wall time: 0:00:01
+```
+{: .log.lammps}
+
+## Useful keyword to search for
+  * The ```Loop time``` is the total wall-clock time for the simulation to run. (source: LAMMPS manual)
+Use the following to extract this from ```log.lammps```:
+```
+grep "Loop time" log.lammps| sed 's/|/ /' | awk '{print $4}'
```
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:187]: This is potentially a really complicated/intimidating command line for people who are not really familiar with bash. While it is convenient, is it really necessary? We can just highlight the number in the output

File: _episodes/02-benchmark-and-scaling.md [hunk 22]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+```
+```
+$ 1.76553
+```
+
+  * The Performance line is provided for convenience to help predict how long it will take to run a desired physical simulation.  (source: LAMMPS manual)
+Use the following command line to extract the value in units of ```tau/day``` :
+```
+grep "Performance" log.lammps| sed 's/|/ /' | awk '{print $2}'
```
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:196]: Again  I would include the line and highlight this information

File: _episodes/02-benchmark-and-scaling.md (hunk 13 above)
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:202]: `1 if...`

File: _episodes/02-benchmark-and-scaling.md [hunk 23]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+```
+```
+$ 24468.549
+```
+
+  * The CPU use line provides the CPU utilization per MPI task; it should be close to 100% times the number of OpenMP threads (or 1 of not using OpenMP). Lower numbers correspond to delays due to file I/O or insufficient thread utilization. (source: LAMMPS manual)
+Use the following command line to extract the value in units of ```tau/day``` :
+```
+grep "CPU use" log.lammps| sed 's/|/ /' | awk '{print $1}'
```
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:205]: Same again

File: _episodes/02-benchmark-and-scaling.md [hunk 24]
```diff
@@ -19,27 +19,244 @@ keypoints:
...
+
+  * Compare this data among various HPC platforms (JSC/Kay/LAMMPS-data): Benchmark plot
+  
 ## Scaling
 
-Good scaling vs poor scaling. How to choose no. of nodes, preventing waste of resources.
+Scaling in computation is the effective use of resources. To explain this in more detail, lets head back to our chefs again from the previous episode
 
-(FIXME)
+Let us assume that they are all bound by secrecy, and are not allowed to reveal to you what their craft is, pastry, meat, fish, soup, etc. You have to find out what their specialities are, what do you do? Do a test run and assign a chef to each course. Having a worker set to each task is all well and good, but there are certain combinations which work and some which do not, you might get away with your starter chef preparing a fish course, or your lamb chef switching to cook beef and vice versa, but you wouldn't put your pastry chef in charge of the main meat dish, you leave that to someone more qualified and better suited to the job. Eventually after a few test meals, you find out the best combination and you apply that to all your future meals.
```
Comment (ocaisa) [inline _episodes/02-benchmark-and-scaling.md:258]: Really fun example!

File: _episodes/05-accelerating-lammps.md [hunk 25]
```diff
@@ -13,6 +13,121 @@ objectives:
 keypoints:
 - "First key point. Brief Answer to questions. (FIXME)"
 ---
+## How can I accelerate LAMMPS performance?
+Well, there are two basic approaches to speed-up LAMMPS. One is to use better algorithm for certain types of calculation, and the other one is to use highly optimized codes via various "accelerator packages" deviced for hardware specific platforms. 
+
+One popular example of first type of approach is to use the Wolf summation method instead of the Ewald summation method for calcultaing long range Coulomb interaction effectively using a short-range potential. Similarly there are a few FFT schemes are offered by LAMMPS and a user has to make a trade-off between accuracy and performance depending on his computational needs. The current tutorial is not aimed to discuss such types of algorithm based speed-up of LAMMPS rather we'll be discussing mainly on a few accelerator packages that is used to extract the most out of an available hardware of a HPC system.
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md:19]: Very nice summary!

File: _episodes/05-accelerating-lammps.md [hunk 26]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+
+There are five accelerator packages currently offered by LAMMPS. These are **OPT**, **USER-INTEL**, **USER-OMP**, **GPU** and **Kokkos**. Specialized codes contained in these packages help LAMMPS to perform well on modern HPC platforms which could have different hardware partitions. Therefore, the very next question that arises that what are these hardwares that are supported by these packages?
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md:21]: Careful, do we know what is meant by `hardware partition here`?

File: _episodes/05-accelerating-lammps.md [hunk 27]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+>
+> | Hardware | Accelarators |
+> | -------- | ------------ |
+> | Multi-core CPUs | OPT, USER-INTEL, USER-OMP, Kokkos |
+> | Intel Xeon Phi | USER-INTEL, Kokkos |
+> | NVIDIA GPU | GPU, Kokkos |
+>
+{: .callout}
+
+Within the limited scope of this tutorial, this is almost impossible to discuss all of the above packages here. The key point to understand here is that the acceleration is achieved by multithreading either through OpenMP or GPU. The ONLY accelerator package that supports both kinds of hardwares is **Kokkos**. Kokkos is a templated C++ library developed in Sandia National Laboratory and this helps to create an abstraction that allows a *single implementation* of a software application on different kinds of hardwraes by simply mapping C++ kernel onto various backend languages.
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md:33]: `different kinds of hardware`

Careful of using too many technical terms here (`templated C++`, `kernel`, `backend`) we can probably come up with an explanation for the layman with resorting to this (and I bet Kokkos already have this on their website)

File: _episodes/05-accelerating-lammps.md [hunk 28]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+> | -------- | ------------ |
+> | Multi-core CPUs | OPT, USER-INTEL, USER-OMP, Kokkos |
+> | Intel Xeon Phi | USER-INTEL, Kokkos |
+> | NVIDIA GPU | GPU, Kokkos |
+>
+{: .callout}
+
+Within the limited scope of this tutorial, this is almost impossible to discuss all of the above packages here. The key point to understand here is that the acceleration is achieved by multithreading either through OpenMP or GPU. The ONLY accelerator package that supports both kinds of hardwares is **Kokkos**. Kokkos is a templated C++ library developed in Sandia National Laboratory and this helps to create an abstraction that allows a *single implementation* of a software application on different kinds of hardwraes by simply mapping C++ kernel onto various backend languages. 
+
+Before discussing on Kokkos, we'll touch a few key points about other accelerator packages to give you a feel about what these packages offer and in many cases these pakckages outperform Kokkos in its current form!
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md:35]: `Before discussing Kokkos`...`these packages`

Comment (ocaisa) [reply _episodes/05-accelerating-lammps.md:35]: Be careful about saying these packages outperform Kokkos, the developers may not like seeing this and it is  the kind of statement that gets people in trouble. As mentioned there are a lot of caveats to performance, e.g., if Kokkos supports something the GPU package does not then Kokkos will always outperform  the GPU package when that is the use case.

File: _episodes/05-accelerating-lammps.md [hunk 29]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+> > ## Solution
+> > The *Pair* component will see a reduction in cost since this accelerator package aims to work on the pair styles only.
+> {: .solution}
+{: .discussion}
+
+> ## USER-INTEL package
+>
+> * Acceleration, in this case, is mainly achieved in two different ways. One of them is to use vectorisation on multi-core CPUs and the other one is to offload calculations of neighbour list and non-bonded interactions to Phi co-processors.
+> 
+> * For using offload feature, we need Intel Xeon Phi coprocessors
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md:70]: maybe mention that these are no longer in production

File: _episodes/05-accelerating-lammps.md [hunk 30]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+> | tersoff               |       |      |        |    |    |     |      |
+> 
+> * This packages supports *single*, *double* and *mixed* precision calculations.
+>
+> * Performance enhancemnt using this package depends on many considerations like the hardware that is available to you, various styles that you are using in the input, size of your problem, and precision. For example, if you are using a pair style (say, reax) for which this is not implemented, its obvious that you are not going to have a performance gain for the *Pair* part of the calculation. Now, if majority of the computation time is coming from the *Pair* part then you are in trouble. If you would like to know how much speedup you can achieve using USER-INTEL, you can look [here](https://lammps.sandia.gov/doc/Speed_intel.html) 
+{: .callout}
+
+> ## USER-OMP package
+>
+> * Acceleration, in this case, is mainly achieved by using templeted C++ library to reduce computational overheads due to if tests and other conditional code blocks.
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md]: `templated`

Comment (ocaisa) [reply _episodes/05-accelerating-lammps.md]: Use
```
`if`
```
 since you mean a programmatic if

File: _episodes/05-accelerating-lammps.md [hunk 31]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+> 
+> * This also provides better vectorization operations as compared to its regular CPU version.
+> 
+> * Only a handful of pair styles can be accelerated using this package. As of *3Mar20* version of LAMMPS, 10 pair styles are supported by this accelerator package. These are ```pair_eam_alloy```, ```pair_eam_fs```, ```pair_eam```, ```pair_lj_charmm_coul_long```, ```pair_lj_cut_coul_long```, ```pair_lj_cut```, ```pair_lj_cut_tip4p_long```, ```pair_lj_long_coul_long```, ```pair_morse``` and ```pair_ufm```.
+> 
+> * This generally offers 5-20% savings on computational cost on most of the machines
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md]: Probably needs a reference

File: _episodes/05-accelerating-lammps.md [hunk 32]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+{: .callout}
+
+> ## GPU package
+>
+> * Acceleration, in this case, is mainly achieved by using templeted C++ library to reduce computational overheads due to if tests and other conditional code blocks.
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md]: `templated` and
```
`if`
```

File: _episodes/05-accelerating-lammps.md [hunk 33]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+> 
+> * This also provides better vectorization operations as compared to its regular CPU version.
+> 
+> * Only a handful of pair styles can be accelerated using this package. As of *3Mar20* version of LAMMPS, 10 pair styles are supported by this accelerator package. These are ```pair_eam_alloy```, ```pair_eam_fs```, ```pair_eam```, ```pair_lj_charmm_coul_long```, ```pair_lj_cut_coul_long```, ```pair_lj_cut```, ```pair_lj_cut_tip4p_long```, ```pair_lj_long_coul_long```, ```pair_morse``` and ```pair_ufm```.
+> 
+> * This generally offers 5-20% savings on computational cost on most of the machines
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md]: Really? Exactly the same?

File: _episodes/05-accelerating-lammps.md [hunk 34]
```diff
@@ -13,6 +13,121 @@ objectives:
...
+
+> ## Kokkos package
+>
+> * Acceleration, in this case, is mainly achieved by using templeted C++ library to reduce computational overheads due to if tests and other conditional code blocks.
+> 
+> * This also provides better vectorization operations as compared to its regular CPU version.
+> 
+> * Only a handful of pair styles can be accelerated using this package. As of *3Mar20* version of LAMMPS, 10 pair styles are supported by this accelerator package. These are ```pair_eam_alloy```, ```pair_eam_fs```, ```pair_eam```, ```pair_lj_charmm_coul_long```, ```pair_lj_cut_coul_long```, ```pair_lj_cut```, ```pair_lj_cut_tip4p_long```, ```pair_lj_long_coul_long```, ```pair_morse``` and ```pair_ufm```.
+> 
+> * This generally offers 5-20% savings on computational cost on most of the machines
```
Comment (ocaisa) [inline _episodes/05-accelerating-lammps.md]: Seems that these are currently copy-pasted

**Review by ocaisa — CHANGES_REQUESTED** [review]

**Comment by cilwerner** [comment]
Still needs reviewing and changing on my part, edits can be made after merge.
